core/state/snapshot/conversion.go

```diff
@@ -144,13 +194,125 @@ func (stat *generateStats) reportDone() {
 		ctx = append(ctx, []interface{}{"slots", stat.slots}...)
 	}
 	ctx = append(ctx, []interface{}{"elapsed", common.PrettyDuration(time.Since(stat.start))}...)
-	log.Info("Generated trie hash from snapshot", ctx...)
+	log.Info("Iterated snapshot", ctx...)
+}
+
+// runReport periodically prints the progress information.
+func runReport(stats *generateStats, stop chan bool) {
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			stats.report()
+			timer.Reset(time.Second * 8)
+		case success := <-stop:
+			if success {
+				stats.reportDone()
+			}
+			return
+		}
+	}
+}
+
+// subTask wraps the necessary information of a task derived by generateTrieRoot.
+type subTask struct {
+	account common.Hash
+	code    common.Hash
+	root    common.Hash
+	call    leafCallbackFn
+	db      ethdb.KeyValueWriter
+	stats   *generateStats
+}
+
+// runSubTasks is a helper function of generateTrieRoot. If generateTrieRoot has
+// the callback for leaves, all sub-tasks can be accumulated and executed here.
+// If any sub-task failed, a signal will be throw back very soon.
+//
+// Note we expect the out channel has at least 1 slot available so that sending
+// won't be blocked.
+func runSubTasks(in chan subTask, out chan error, stop chan struct{}) {
+	var (
+		running int
+		failed  bool
+		limit   = runtime.NumCPU()
+		done    = make(chan error)
+	)
+	if cap(out) < 1 {
+		panic("require buffered channel")
+	}
+	run := func(task subTask) {
+		root, err := task.call(task.db, task.account, task.code, task.stats)
+		if err != nil {
+			done <- err
+			return
+		}
+		if task.root != root {
+			done <- fmt.Errorf("invalid subroot(%x), want %x, got %x", task.account, task.root, root)
+			return
+		}
+		done <- nil
+	}
+	schedule := func(task subTask) {
+		// Short circuit if failure already occurs.
+		if failed {
+			return
+		}
+		running += 1
+		go run(task)
+
+		// If there are too many runners, block here
+		for running >= limit {
+			failure := <-done
+			running -= 1
+			if failure != nil && !failed {
+				failed = true
+				out <- failure // won't be blocked
+			}
+		}
+	}
+	for {
+		select {
+		case task := <-in:
+			schedule(task)
+		case failure := <-done:
+			running -= 1
+			if failure != nil && !failed {
+				failed = true
+				out <- failure // won't be blocked
```

Comment:
I don't see why this could not block. The `out` channel has a buffer of `1`. If I assign 3 subtasks and all 3 fail more or less at the same time (e.g. disk io error, db closed, etc), I'll end up with 3 errors on the `done` channel simultaneously. Then the first will be delivered into `out`, when that's read by `generateTrieRoot` the second gets delivered too, which will never be read. The third failure will deadlock.

Reply:
`failed` will be marked for any failure. And all following failures are just ignored.